core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts

```diff
@@ -235,20 +230,32 @@ export class DiagnosticTestPlayerEngineService {
   }
 
   getRecommendedTopicIds(): string[] {
+    // The method used to recommend topics to the learner based on their
+    // performance in the diagnostic test.
     let recommendedTopicIds: string[] = [];
     let failedTopicIds: string[] = this.getFailedTopicIds();
+    // The topics which do not contain any prerequisites are referred to as
+    // root topics.
     let rootTopicIds: string[] = this.getRootTopicIds();
     let failedRootTopicIds: string[] = rootTopicIds.filter(
       topicId => failedTopicIds.indexOf(topicId) !== -1
     );
 
     if (failedRootTopicIds.length >= 2) {
+      // If the learner failed in two or more root topics, then any two root
+      // topics were recommended, so that learner can start with any
+      // one of them.
       recommendedTopicIds.push(failedRootTopicIds[0]);
       recommendedTopicIds.push(failedRootTopicIds[1]);
     } else if (failedRootTopicIds.length === 1) {
+      // Among the failed topics, if any of the topics is a root topic, then
+      // it must be recommended.
       recommendedTopicIds.push(failedRootTopicIds[0]);
     } else {
-      let sortedTopicIds = this.getSortedTopicIds();
+      // If none of the failed topics is a root topic, then the failed topics
+      // are topologically sorted and then the first topic among the sorted list
+      // will be recommended.
+      let sortedTopicIds = this.getSortedInitialTopicIds();
```

Comment:
I don't know what "sorted initial topic IDs" means. Couldn't you just say getTopologicallySortedTopicIds()?

But also, I'm not sure which topic IDs you are referring to -- how does this method know which topics are failed? Shouldn't we pass in a list of failed topic IDs to it so that it can topologically sort those? The word "initial" doesn't convey anything to me, and if anything it sounds like the full list of topics which seems different from just the failed ones (as per the comment).

Lastly, I'm not even sure a full top-sort is needed -- couldn't we just compute a list of topic IDs which have no prerequisites within the set?

Reply:
Updated the name.

This method only sorts the topics. This method will be used internally by the getTopicRecommendations() function.
The recommendation logic:
- Check if the failed topics have any root topics.
    - If 2 or more roots are present in the failed topics: Recommend any two roots.
    - If 1 root topic is failed: Recommend the root topic.
- If no root topics then,
    - Sort the topic IDs that were initially present i.e., topicIdToPrerequisiteTopicIds. Here I am sorting all the topics not only the failed ones because the failed topics may or may not have dependencies with each other.
    - After sorting I am picking the first topic from the sorted list that is failed by the learner.

Answering the last point: Can you please confirm that in the last para, are you referring to the root topic IDs (i.e., compute a list of root topics, because they have no prerequisites)?
If yes: The failed topic IDs may or may not contains the root topics, if they contain any root topic, then the recommendation function will take care of that. Also, we only need to sort the topics if the failed topics list does not contain any of the root topics.
If not: Please help me to understand the last point.

Thank you

Reply:
Hi @Nik-09, I almost missed this! Please remember to reassign to whoever needs to take the next action, even though it might just be answering a question.

For the last point, I was talking about looking for topics within the failed topics that do not have prerequisites _within the failed topics_. I.e. consider the nodes that consist of the _failed topics only_. Within that graph, we just want to return the topics that have no other prerequisites, right?

E.g. if you have A --> B, B --> C, A --> D, D --> C, C --> E and only B, D, C, E are failed, then B and D are root topics in the "failed graph" and the ones that would be recommended. C has prerequisites B/D and E has prerequisite C, so they're not eligible. Although B and D have A as a prerequisite, A isn't a failed topic and isn't in the smaller "failed graph".

Does this clarify?

Reply:
Yes, I got the point, but originally we have a dict containing all the topicIdToPrerequisite topic IDs.

In the end, when we have to sort the failed topic ID, we just need a dependency graph for that we have to remove the non-failed topics from the original dependency graph, which may be extra.

Now in order to minimize the full topological sorting I think terminating the function if we encounter the first failed ID while traversing the full graph may improve the efficiency.

So based on that, the new method name can be: getTopologicallyFirstTopicId()
The method returns the first topic ID from the topologically sorted list of the given topic IDs.

Can you please provide feedback on this approach?

Reply:
I think something like that is totally fine. Not sure if you want to return 1 or 2 IDs but the basic idea makes sense to me!